content/blog/2025-12-16-facade-layout-compiler.md

```diff
@@ -0,0 +1,409 @@
+
+---
+
+## What were the hardest parts to get right?
+
+### 1. Separating Syntax Errors from Semantic Errors
+
+The distinction between lexical, syntactic, and semantic errors was initially blurry. For example:
+
+- Is `@` an invalid character (lexical) or an invalid symbol (semantic)?
+- Is `row 1: X E E` a syntax error or a semantic error?
+
+I resolved this by establishing boundaries:
+
+- **Lexical errors:** Characters not in the DSL's alphabet (punctuation, Unicode, emoji)
+- **Syntax errors:** Malformed constructs (missing colons, invalid repeat expressions, missing row numbers)
+- **Semantic errors:** Valid syntax with invalid meaning (floating doors, disconnected chimneys, invalid symbol letters like `X`)
+
+The key insight was that `X` is a valid identifier syntactically but not a valid symbol semantically. The parser accepts it and emits a semantic error during cell processing.
+
+### 2. Chimney Connectivity Analysis
+
+Initially, I tried a simple positional check (e.g., chimney cells must be on the top row). This failed for common patterns like diagonal chimneys:
+
+```
+row 1: C E E
+row 2: C C E
+row 3: E C C
+row 4: E E C
+```
+
+This chimney is valid. It connects to the roof through a diagonal path. The flood-fill approach correctly handles this by treating chimneys as connected components rather than isolated cells.
+
+A subtle bug emerged with multiple chimney components: the algorithm must check each component independently, not just verify that some chimney reaches the roof.
+
+### 3. Zero-Count Repeat Expressions
+
+The expression `E*0` presented an interesting design decision. Options:
+
+1. Treat as syntax error (reject)
+2. Treat as single cell (emit one `E`)
+3. Treat as zero cells (emit nothing)
+
+I chose option 3 because it's the most mathematically consistent interpretation and enables useful patterns like conditional exclusion. The implementation simply doesn't add any cell to the AST when count is 0.
+
+### 4. Normalization Without Hiding Bugs
+
+Auto-filling rows and padding widths improved usability but introduced a risk: it could mask user errors. If a user forgets a cell, the compiler silently adds an empty cell, which potentially hiding bugs.
+
+I addressed this by:
+
+- Explicitly tracking auto-filled cells in metadata
+- Emitting INFO messages about auto-fill actions
+- Including auto-fill information in JSON output
+
+This makes compiler intervention visible without treating it as an error.
+
+### 5. Error Recovery and Multiple Error Reporting
+
+Fixing one error often revealed another, requiring multiple compile-fix cycles.
+
+I implemented error recovery in both the lexer and parser:
+
+- **Lexer:** Reports invalid characters but continues scanning
+- **Parser:** Uses `sync_to_row()` to skip to the next row declaration after errors
+
+This required careful state management to ensure the compiler doesn't report spurious follow-on errors while still catching genuine independent errors.
```

Comment:
What does that "careful state management" consist of? What kind of state are you referring to?